torch/csrc/jit/codegen/cuda/compute_at_map.cpp

```diff
@@ -931,130 +869,233 @@ IterDomain* ComputeAtMap::computeConcreteId(
+  int max_iter_root_count = 0;
+  int max_bcast_root_count = 0;
+
+  for (auto maybe_concrete_id : maybe_concrete_ids.vector()) {
+    auto concrete_id_root_sets = getInputDisjointSetsOf(maybe_concrete_id);
 
     int bcast_root_count = std::count_if(
-        root_ids.begin(), root_ids.end(), [&](IterDomain* root_id) {
-          return root_id->isBroadcast()
-              // TODO: This shouldn't have a negative impact, but (emperically)
-              // might not be necessary
-              || trivial_reduction_info_.isDerived(root_id);
+        concrete_id_root_sets.vector().begin(),
+        concrete_id_root_sets.vector().end(),
+        [&](std::shared_ptr<VectorOfUniqueEntries<IterDomain*>> set) {
+          return set->vector()[0]->isBroadcast() ||
+              set->vector()[0]->isTrivialReduction() ||
+              trivial_reduction_info_.isDerived(set->vector()[0]);
         });
-    int iter_root_count = (int)root_ids.size() - bcast_root_count;
+
+    int iter_root_count =
+        (int)concrete_id_root_sets.vector().size() - bcast_root_count;
     if (iter_root_count > max_iter_root_count ||
         (iter_root_count == max_iter_root_count &&
          bcast_root_count > max_bcast_root_count)) {
       max_iter_root_count = iter_root_count;
       max_bcast_root_count = bcast_root_count;
       concrete_id = maybe_concrete_id;
-
-      // If we update the concrete_id, then update the root_ids_of_concrete_id
-      // to reflect this id
-      if (mode == IdMappingMode::LOOP) {
-        root_ids_of_concrete_id.clear();
-        std::transform(
-            root_ids.begin(),
-            root_ids.end(),
-            std::inserter(
-                root_ids_of_concrete_id, root_ids_of_concrete_id.end()),
-            [&](const auto root_id) {
-              return getConcreteMappedID(root_id, IdMappingMode::EXACT);
-            });
-      }
     }
-  } // end maybe_concrete_id
+  }
 
   TORCH_INTERNAL_ASSERT(
       concrete_id != nullptr,
-      "Something went wrong, could not find a concrete id.");
-
-  if (mode == IdMappingMode::LOOP) {
-    // Validate the concrete id has influence from all the roots of all the
-    // consumers that will map to this concete id in the loop map. This means
-    // all the consumers in all expressions of the loop nest generated based on
-    // this concrete ID will have their roots mapping to this concrete ID
-    // represented in the extent of this concrete id.
-    validateCompletenessOfLoopConcreteID(
-        concrete_id,
-        *this,
-        trivial_reduction_info_,
-        root_ids_of_all_ids,
-        root_ids_of_concrete_id,
-        root_id_to_maybe_concrete_ids,
-        disjoint_set_shared_ptr->vector(),
-        maybe_concrete_ids.vector());
-  }
+      "No concrete_id found for disjoint set ",
+      disjoint_set_shared_ptr->toString());
 
   return concrete_id;
 }
 
 void ComputeAtMap::buildConcreteIds() {
+  for (const auto& disjoint_set_shared_ptr :
```

Comment:
Would be helpful to have some comment on why only the exact concrete IDs are chosen by just picking the first IDs.

Reply:
Yes, you're right. Will add a comment.